---
title: "CRISPR"
subtitle: "Final Project: Statistics C245F"
author: "Florica Constantine, Lingrong Jin, Andy Shen"
header-includes:
   - \usepackage{float}
   - \usepackage{bbm}
output: 
  pdf_document:
    number_sections: true
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
rm(list = ls())
library(tidyverse)
library(GGally)
library(corrplot)
library(cvms)
library(ggpubr)
library(reshape)
library(limma)
library(enrichR)
library(patchwork)
library(ggVennDiagram)
library(MASS)
library(kableExtra)
library(knitr)
knitr::opts_chunk$set(
  echo = FALSE,  # don't print the code chunk
  warning = FALSE,  # don't print warnings
  message = FALSE,  # don't print messages
  fig.width = 6,  # set default width of figures
  fig.height = 4,  # set default height of figures
  fig.align = "center",  # always align figure in center
  fig.pos = "H",  # always plot figure at the exact location of the code chunk
  cache = FALSE) 
setwd("/Users/andyshen/Library/Mobile Documents/com~apple~CloudDocs/Desktop/Desktop - Andyâ€™s MacBook Pro/UC Berkeley/spring-2022/stat-c245f/stat-c245f-hw2")
```

\newcommand{\blue}{\textcolor{blue}}
\newcommand{\red}{\textcolor{red}}
\newcommand{\bb}{\textbf}


```{r}
# read in data
count_data <- read.csv('./Count_output/Comparison_to_5k_plasmid.count_normalized.txt', sep='\t')
count_output <- read.csv('./Count_output/Comparison_to_5k_plasmid.count.txt', sep='\t')
count_matrix <- count_output[,3:9]
cor <- cor(count_matrix)
col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))
count_matrix_log <- log(count_output[,3:9]+0.00000001)
cor_log <-cor(count_matrix_log)


mcd <- readRDS("mcd_df.rds")
mh <- readRDS("mh_df.rds")
tmcd <- readRDS("tmcd_df.rds")
tmh <- readRDS("tmh_df.rds")
```

```{r}
path_beaddp <- "./Test_outputs/Bead_DP_v_plasmid5k/"
path_ccdp <- "./Test_outputs/Cc_DP_v_plasmid5k/"
path_beaddn <- "./Test_outputs/Bead_DN_v_plasmid5k/"
path_ccdn <- "./Test_outputs/Cc_DN_v_plasmid5k/"

bead_dp <- read.table(paste0(path_beaddp,"Comparison_BeadDP_to_5k_plasmid_output.gene_summary.txt"), 
                       header = TRUE)
bead_dn <- read.table(paste0(path_beaddn,"Comparison_BeadDN_to_5k_plasmid_output.gene_summary.txt"), 
                       header = TRUE)
cc_dp <- read.table(paste0(path_ccdp,"Comparison_CcDP_to_5k_plasmid_output.gene_summary.txt"), 
                       header = TRUE)
cc_dn <- read.table(paste0(path_ccdn,"Comparison_CcDN_to_5k_plasmid_output.gene_summary.txt"), 
                       header = TRUE)
```




# Introduction

The discovery of the Cas9 clustered regularly interspaced short palindromic repeats (CRISPR) enzyme by @ran2013genome has revolutionized the world of genomics and bioinformatics. The ability to perform genome editing and cleaving opens the door to an endless array of biological research that explores the effect of Cas9.

One method is the Model-based Analysis of Genome-wide CRISPR/Cas9 Knockout (MAGeCK) proposed by @li2014mageck. This technique utilizes the unique ability of Cas9 by identifying both positively and negatively-regulated genes in a robust and powerful manner. Here we utilize the output of MAGeCK on a CRISPR assay for children with Severe combined immunodeficiency (SCID). This rare disorder directly inhibits T-cells from being created, significantly altering a child's immune system. Such a disorder is typically fatal^[https://www.niaid.nih.gov/diseases-conditions/severe-combined-immunodeficiency-scid].

However, the advent of CRISPR assay sequencing technologies enables domain experts to investigate the genomic origins of SCID and discover genes that play a role in immune system development. In this process, a guide RNA (sgRNA) enters the genome with the intention to target a gene. A dissipating guide indicates that the sgRNA, and hence its target gene, was important for the specific cell-line. These discoveries point out the potential usefulness of certain genes in T-cell differentiation.

# Data Organization

Our data is presented as a series of counts per gene. These counts correspond to the number of cells observed per gene. These cells have markers which indicate the presence of T-cells. A double-negative (DN) count indicates the presence of a cell with no T-cell markers. A single-positive (SP) count indicates one T-cell marker, while a double-positive (DP) count indicates the presence of two T-cell markers, the clearest evidence of T-cell detection. 

While each sgRNA has a fixed efficiency, we are not provided that information, so this analysis assumes that the efficiency is fixed across all sgRNAs.

# Exploratory Data Analysis

We compute the column-wise correlation between all columns in our data set in order to get an initial sense on how our data is distributed. These values are summarized in Figure \blue{\ref{fig:colcorr}}. We observe that that there exists a moderately strong association between some columns, namely between SP columns and other columns. These associations justify using the joint distribution of the columns as opposed to pairwise column distributions.

```{r colcorr, fig.height=5, fig.width=7, fig.cap="Correlation of all columns with each other in nornalized data set.", out.width='85%'}
corrplot(cor_log, type = "upper", method = "number")
```





# Inhibitors of T-Cell Differentiation

```{r}
num_guide_df <- bead_dn
num_guide_df$Gene <- num_guide_df$id
full_counts_norm <- full_join(count_data, num_guide_df, by = "Gene")
```




```{r}
# determining inhibitors vs promoters
extractInhibitors <- function(data) {
  cols <- c("sgRNA", "Gene", "BeadDP", "BeadDN", "CcDP" , "CcDN" , "Plasmid5kCTRL")
  cand_genes <- data$gene
  sap_out <- sapply(cand_genes, function(g) {
    cdf <- full_counts_norm %>% filter(Gene == g) %>% 
    dplyr::select(all_of(cols)) %>% 
    mutate(across(c(BeadDN, BeadDP, CcDP, CcDN), ~ .x - Plasmid5kCTRL)) %>% 
    dplyr::select(-Plasmid5kCTRL) %>% 
    summarise(across(where(is.numeric), mean))
    inhib_bead <- ifelse(cdf$BeadDP <= cdf$BeadDN, "Inhibitor", "Promoter")
    inhib_cc <- ifelse(cdf$CcDP <= cdf$CcDN, "Inhibitor", "Promoter")
    out <- c(inhib_bead, inhib_cc)
    names(out) <- c("Bead", "Cc")
    out
  })
  t(sap_out) %>% data.frame()
}
```



```{r}
mcd_inhib <- extractInhibitors(mcd)
mh_inhib <- extractInhibitors(mh)
tmcd_inhib <- extractInhibitors(tmcd)
tmh_inhib <- extractInhibitors(tmh)
```


```{r, eval=FALSE}
saveRDS(mcd_inhib, "./inhibitors/mcd_inhib.rds")
saveRDS(mh_inhib, "./inhibitors/mh_inhib.rds")
saveRDS(tmcd_inhib, "./inhibitors/tmcd_inhib.rds")
saveRDS(tmh_inhib, "./inhibitors/tmh_inhib.rds")
```

```{r}
both_inhib <- mcd_inhib %>% data.frame() %>% 
  mutate(full_inhib = ifelse(Bead == "Inhibitor" & Cc == "Inhibitor", TRUE, FALSE)) %>% 
  filter(full_inhib == TRUE)
# both are inhibitor
```



In order to determine whether the genes we discovered using these four methods are inhibitors of T-cell differentiation or not, we compare the average difference in DN and DP T-cell counts with respect to the control. However, since each gene has a different number if guide RNAs assigned to it, we must scale the counts by the number of guide RNAs assigned to each gene. If the average difference of DN counts with the control is larger than that of DP with the control, we can infer the gene as an inhibitor, otherwise it is a promoter. 


We discover 9 genes that are classified as inhibitor by both the Bead and cell co-culture. These genes are `r rownames(both_inhib)`. The biological significance of these genes is explained further in Section \blue{\ref{sec:enrichr}}.

\red{Do we need to write more here?}

To further compare our results with those of MAGeCK, we examine the overlapping significant genes between the two methods. Figure \blue{\ref{fig:vds}} shows the number of genes that overlap between the two methods. Note that for the MAGeCK algorithm, we only consider genes that were selected with negative p-value less than 0.05.

```{r}
bead_dp_sig <- bead_dp %>% filter(neg.p.value < 0.05)
cc_dp_sig <- cc_dp %>% filter(neg.p.value < 0.05)
```

```{r vds, fig.cap="Overlap of discovered genes from MAGeCK (Bead/Cc DP) and our MCD method."}
vd_bdp_mcd <- ggVennDiagram(list("Bead DP" = bead_dp_sig$id, "MCD" = mcd$gene)) + 
  theme(legend.position = "none")
vd_ccdp_mcd <- ggVennDiagram(list("Cc DP" = cc_dp_sig$id, "MCD" = mcd$gene)) + 
  scale_fill_gradient(low = "#F6BDC0", high = "#DC1C13") + 
  theme(legend.position = "none")
vd_bdp_mcd + vd_ccdp_mcd
```

These results indicate that MCD is especially amenable to the bead genes as 21 genes overlap between the two methods. Moreover, MCD is slightly more conservative than MAGeCK since we selected fewer genes. While our results show strong overlap, future work for corroborating the efficacy of these methods requires an in-depth look into the selected genes and the role they play in T-cell differentiation.  

# EnrichR {#sec:enrichr}

```{r}
dir <- getwd()
bead_genes_DP <- read.table(paste0(dir,"/Test_outputs/Bead_DP_v_plasmid5k/Comparison_BeadDP_to_5k_plasmid_output.gene_summary.txt"), header = T)
cc_genes_DP <- read.table(paste0(dir,"/Test_outputs/Cc_DP_v_plasmid5k/Comparison_CcDP_to_5k_plasmid_output.gene_summary.txt"), header = T)

bead_genes_SP <- read.table(paste0(dir,"/Test_outputs/Bead_SP_v_plasmid5k/Comparison_BeadSP_to_5k_plasmid_output.gene_summary.txt"), header = T)
cc_genes_SP <- read.table(paste0(dir,"/Test_outputs/Cc_SP_v_plasmid5k/Comparison_CcSP_to_5k_plasmid_output.gene_summary.txt"), header = T)

bead_genes_DN <- read.table(paste0(dir,"/Test_outputs/Bead_DN_v_plasmid5k/Comparison_BeadDN_to_5k_plasmid_output.gene_summary.txt"), header = T)
cc_genes_DN <- read.table(paste0(dir,"/Test_outputs/Cc_DN_v_plasmid5k/Comparison_CcDN_to_5k_plasmid_output.gene_summary.txt"), header = T)
```

```{r}
# inhibitors predicted by mageck include negatively selected DP and positively selected DN
#DP
bead_DP_sig <- bead_genes_DP %>% 
  arrange(neg.rank) %>%
  filter(neg.p.value<0.05)%>%
  dplyr::select(id)
cc_DP_sig <- cc_genes_DP %>% 
  arrange(neg.rank) %>%
  filter(neg.p.value<0.05)%>%
  dplyr::select(id)
bead_DP_sig$method <- rep('bead',nrow(bead_DP_sig))
cc_DP_sig$method <- rep('cc',nrow(cc_DP_sig))
bead_DP_sig$response <- rep('DP',nrow(bead_DP_sig))
cc_DP_sig$response <- rep('DP',nrow(cc_DP_sig))

#DN
bead_DN_sig <- bead_genes_DN %>% 
  arrange(pos.rank) %>%
  filter(pos.p.value<0.05)%>%
  dplyr::select(id)
cc_DN_sig <- cc_genes_DN %>% 
  arrange(pos.rank) %>%
  filter(pos.p.value<0.05)%>%
  dplyr::select(id)
bead_DN_sig$method <- rep('bead',nrow(bead_DN_sig))
cc_DN_sig$method <- rep('cc',nrow(cc_DN_sig))
bead_DN_sig$response <- rep('DN',nrow(bead_DN_sig))
cc_DN_sig$response <- rep('DN',nrow(cc_DN_sig))

mageck_inhib <- bind_rows(bead_DP_sig,cc_DP_sig,bead_DN_sig,cc_DN_sig)

mageck_inhib <- distinct(mageck_inhib,id,.keep_all = T) #253
```

```{r}
#mcd_inhib <- readRDS("~/stat-c245f-hw2/inhibitors/mcd_inhib.rds")
mcd_inhib <- readRDS("./inhibitors/mcd_inhib.rds")
mcd_inhib$gene <- rownames(mcd_inhib)
rownames(mcd_inhib) <- NULL

# mcd_inhib_bead <- mcd_inhib %>%
#   filter(Bead == "Inhibitor") #38
```


```{r}
mageck_inhib_bead <- mageck_inhib %>%
  filter(method == "bead" & response == "DP") #69
websiteLive <- TRUE
dbs <- c("GO_Molecular_Function_2021", "GO_Cellular_Component_2021", "GO_Biological_Process_2021", "Human_Phenotype_Ontology")
```


```{r}
# if (websiteLive) {
#     enriched1 <- enrichr(mageck_inhib_bead$id, dbs)
#     saveRDS(enriched1, "enriched1.rds")
# }
```


```{r}
enriched1 <- readRDS("enriched1.rds")
p1 <- if (websiteLive) plotEnrich(enriched1[[1]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Molecular_Function")
p2 <- if (websiteLive) plotEnrich(enriched1[[2]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Cellular_Component")
p3 <- if (websiteLive) plotEnrich(enriched1[[3]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Biological_Process")
p4 <- if (websiteLive) plotEnrich(enriched1[[4]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "Human_Phenotype_Ontology (Bead)")
# p1;p2;p3;p4
```


```{r, message=FALSE}
# if (websiteLive) {
#     enriched2 <- enrichr(mcd_inhib$gene, dbs)
#     saveRDS(enriched2, "enriched2.rds")
# }
```


```{r, message=FALSE}
enriched2 <- readRDS("enriched2.rds")
p5 <- if (websiteLive) plotEnrich(enriched2[[1]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Molecular_Function")
p6 <- if (websiteLive) plotEnrich(enriched2[[2]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Cellular_Component")
p7 <- if (websiteLive) plotEnrich(enriched2[[3]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "GO_Biological_Process")
p8 <- if (websiteLive) plotEnrich(enriched2[[4]], showTerms = 10, numChar = 40, y = "Count", orderBy = "P.value", title = "Human_Phenotype_Ontology")

# p5;p6;p7;p8
```


We performed an enrichment analysis of the essential genes predicted by MCD and by MAGeCK (DP genes with a negative p.value <  0.05) using `enrichR`, an online gene enrichment tool created by @kuleshov2016enrichr. `enrichR` accepts a list of genes and compares them against existing annotated gene sets. This provides domain knowledge for our genes and allows us to assess whether our genes have biological meaning relative to T-cell differentiation. 


As shown in Figure \blue{\ref{fig:enrichr}} below, biological processes seemingly relevant to T cell development appear as significantly enriched for both sets of gene predictions. These processes include T cell receptor signaling and regulation of stem cell differential. 

Similarly, disease phenotypes relevant to T cell malfunction such as T lymphocytopenia and abnormality of T cells appear at the top of enriched terms for both sets of genes. While some enriched terms for genes predicted by MCD seem less relevant compared to those for MAGeCK, both methods are able to discover the essential genes reasonably well.


```{r enrichr, fig.width=12, fig.height=8, out.width='100%', fig.cap="Biological pathways and human phenotypes of genes selected by MAGeCK (top panels) and MCD (bottom panel)."}
mageck_inhib_graph <- ggarrange(p3,p4, ncol =2)
mageck_inhib_graph <- annotate_figure(mageck_inhib_graph, left = text_grob("MAGeCK", rot = 90))
mcd_inhib_graph <- ggarrange(p7,p8, ncol =2)
mcd_inhib_graph <- annotate_figure(mcd_inhib_graph, left = text_grob("MCD", rot = 90))

fig <- ggarrange(mageck_inhib_graph,mcd_inhib_graph, nrow = 2)
fig
```



# References












